### DISCUSSION TITLE
Can we contact two pdf files side by side

### DISCUSSION THREAD

**Discussion by kvrameshreddy** [pymupdf/PyMuPDF#1383] [Q&A] [answered]
Hi @JorjMcKie ,

Is there any option to show two pdf files side by side as shown in the attachment. 
[output.pdf](https://github.com/pymupdf/PyMuPDF/files/7510383/output.pdf)

**Comment by JorjMcKie** [comment]
Sure, use `page.show_pdf_page()`.

Using a few simplifying assumptions:
* pages in both files are of equal dimension, and protrait (take ISO A4 as an example)
* equal number of pages in both files

... the following snippet will do the job

```python
w, h = fitz.paper_size("a3-l")  # width, height of output page (format ISO A3 landscape)
rleft = fitz.paper_rect("a4")  # rectangle of left output page half (A4 portrait)
rright = rleft + (rleft.width, 0, rleft.width, 0)  # the right side rectangle
src1 = fitz.open("file1.pdf")  # contains pages for the left side
src2 = fitz.open("file2.pdf")  # contains page sfor the right side
n = len(src1)  # number of input pages in both files
doc = fitz.open()  # output file
for i in range(n):
    page = doc.new_page(width=w, height=h)  # make new output page, A3-landscape
    page.show_pdf_page(rleft, src1, n)  # put input page on left output page part
    page.show_pdf_page(rright, src2, n)
doc.save("side-by-side.pdf")
```

With a few more or less obvious adaptions, you can also handle cases where input files have unequal page numbers, or to dynamically adapt the output page format, etc.

You may also want to look at [this](https://github.com/pymupdf/PyMuPDF/discussions/1378) discussion to see how an input file can be reformatted to show two pages in a row side-by-side on output ("2-up" formatting).

**Reply by kvrameshreddy** [reply]
Thanks for the solution, 
I am able to join the two pdfs side by side. 
Can we do the same process on the fly on the previously opened files, where we have annotations.

**Comment by JorjMcKie** [comment, marked as answer]
>  Can we do the same process on the fly on the previously opened files, where we have annotations.

I think I don't understand: which previously opened files?

Annotations cannot be shown by this technique. Remember that annotations are technically not a (permanent) part of a PDF. An annotation is valid for the PDF page on which it is shown. So input annots would have to be re-inserted (using diefferent coordinates etc., etc.) in the newly created, large target page. There is no support for this.
But in many cases, you can convert a PDF such that annotations are converted to normal page content. Then the snippet above also works for the conversion result.
Here is how to do the conversion:
```python
pdfbytes = src1.convert_to_pdf()  # doc is any document type, need not be PDF
newpdf1 = fitz.open("pdf", pdfbytes)  # the document (re-) converted to PDF
# this newpdf will no longer have annotations, technically, but show them as normal, permanent content
```
Whether this works or not, depends on the fonts used in the PDF: not all font types are supported for a PDF-to-PDF conversion.

Another approach is to make temporary page images and show **_these_** side-by-side. This would roughly look like this:
```python
...
dpi = 300  # choose quality of the page images
mat = fitz.Matrix(dpi/72, dpi/72)  # make a zoom mattrix
for i in range(n):
    page = doc.new_page(width=w, height=h)
    pageleft = src1[i]  # page for left image
    pageright = src2[i]  # page for right image
    pixleft = pageleft.get_pixmap(matrix=mat)  # make images
    pixright = pageright.get_pixmap(matrix=mat)
    page.insert_image(rleft, pixmap=pixleft)
    page.insert_image(rright, pixmap=pixright)
...
```

**Reply by JorjMcKie** [reply]
Answered your question?

**Reply by kvrameshreddy** [reply]
Thanks @JorjMcKie, the insert_image worked for me. but the generated pdf size is too large.
the input file size is 10kb the newly generated pdf size is 135MB for dpi=300 and if dpi is none the size is 8MB. is there any chance to reduce the size of new pdf

**Comment by JorjMcKie** [comment]
When saving make sure you use garbage>=3 and deflate=True. Also experiment with the dpi.
Have you tried converting the source files as I described?
There is no other alternative available.

**Reply by kvrameshreddy** [reply]
I will try these options, the source file conversion didn't worked for me.

when i converted source file only single annotation appeared in the new pdf.

**Comment by JorjMcKie** [comment]
what you could also try:
For each annotation on an input page, make a picture of the annotation rectangle: `pix = annot.get_pixmap(...)`.
Then compute the rectangle of that image in the target page:
* for the left side, this should be the same rectangle
* for the right side, add (rleft.width, 0, rleft.widt, 0) to the source annotation rectangle
* Then
```python
Page.insert_image(annotrect, pixmap=annot_pixmap)
```
for all annotation pxmaps.
Fat chance, that the overall result is smaller that full page pixmaps.

**Reply by kvrameshreddy** [reply]
ok @JorjMcKie , I will try that. the garbage=3 solved the file size issue also. thankyou very much.

i am marking this as the answer.

**Reply by kvrameshreddy** [reply]
Hi @JorjMcKie , I need small help from you, can we draw a line between the two pdf's at center while concatenating using insert image

**Reply by kvrameshreddy** [reply]
I tried by drawing line like this
page.draw_line((595,0),(595,842),color=(1,0,0))

but i am loosing the line at center of the page(top and bottom of the page line is visible), is getting overlaped by the second page. is there any other way to do this ??

**Reply by JorjMcKie** [reply]
it depends on the point in time you do this:
do it **_after_** you fill the target page with the source pages. This way it will come on top of things on the page already. To make sure you see what is happening, choose a large line width at first, e.g. width=5
